Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithmJArbitraryInTime.cpp

```diff
@@ -0,0 +1,531 @@
+#if defined(WARPX_DIM_3D)
+        const amrex::Real* ky_s = modified_ky_vec[mfi].dataPtr();
+#endif
+        const amrex::Real* kz_s = modified_kz_vec[mfi].dataPtr();
+
+        const amrex::Array4<amrex::Real const> C = C_coef[mfi].array();
+        const amrex::Array4<amrex::Real const> S_ck = S_ck_coef[mfi].array();
+
+        const amrex::Array4<amrex::Real> X5 = X5_coef[mfi].array();
+        const amrex::Array4<amrex::Real> X6 = X6_coef[mfi].array();
+        const amrex::Array4<amrex::Real> X7 = X7_coef[mfi].array();
+
+        // Loop over indices within one box
+        amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept
+        {
+            // Calculate norm of k vector
+            const amrex::Real knorm_s = std::sqrt(
+                amrex::Math::powi<2>(kx_s[i]) +
+#if defined(WARPX_DIM_3D)
+                amrex::Math::powi<2>(ky_s[j]) + amrex::Math::powi<2>(kz_s[k]));
+#else
+                amrex::Math::powi<2>(kz_s[j]));
+#endif
+            // Physical constants and imaginary unit
+            constexpr amrex::Real c = PhysConst::c;
+            constexpr amrex::Real ep0 = PhysConst::ep0;
+
+            // Auxiliary coefficients
+            const amrex::Real dt2 = dt * dt;
+            const amrex::Real dt3 = dt * dt * dt;
+
+            const amrex::Real om_s  = c * knorm_s;
+            const amrex::Real om2_s = om_s * om_s;
+            const amrex::Real om3_s = om2_s * om_s;
+            const amrex::Real om4_s = om2_s * om2_s;
+
+            if (om_s != 0.)
+            {
+                X5(i,j,k) = (dt3*om3_s - 3._rt*dt2*om3_s*S_ck(i,j,k) - 12._rt*dt*om_s*(1._rt+C(i,j,k)) + 24._rt*om_s*S_ck(i,j,k)) / (6._rt*ep0*dt2*om4_s*om_s);
+            }
+            else
+            {
+                X5(i,j,k) = dt3 / (30._rt * ep0);
+            }
+
+            if (om_s != 0.)
+            {
+                X6(i,j,k) =  ( dt*om2_s*S_ck(i,j,k) + 2._rt*C(i,j,k) - 2._rt) / (2._rt*ep0*dt*om4_s);
+            }
+            else
+            {
+                X6(i,j,k) = - dt3 / (24._rt * ep0);
+            }
+            if (om_s != 0.)
+            {
+                X7(i,j,k) =  ( dt - S_ck(i,j,k) ) / (ep0*om2_s);
+            }
+            else
+            {
+                X7(i,j,k) = dt3 / (6._rt * ep0);
+            }
+        });
+    }
+}
+
+void PsatdAlgorithmJArbitraryInTime::CurrentCorrection (SpectralFieldData& field_data)
+{
+    // Profiling
+    BL_PROFILE("PsatdAlgorithmJArbitraryInTime::CurrentCorrection");
+
+    amrex::ignore_unused(field_data);
+    WARPX_ABORT_WITH_MESSAGE(
+        "Current correction not implemented for multi-J PSATD algorithm");
+}
+
+void
+PsatdAlgorithmJArbitraryInTime::VayDeposition (SpectralFieldData& field_data)
+{
+    // Profiling
+    BL_PROFILE("PsatdAlgorithmJArbitraryInTime::VayDeposition()");
+
+    amrex::ignore_unused(field_data);
+    WARPX_ABORT_WITH_MESSAGE(
+        "Vay deposition not implemented for multi-J PSATD algorithm");
+}
```

Comment:
The implementation of these two functions, available only for the case PSATD-CLm, that is, J constant and rho linear, can be copied from the file PsatdAlgorithmSecondOrder.cpp in #3506.

Reply:
Done in [ce502d0](https://github.com/ECP-WarpX/WarpX/pull/4263/commits/ce502d0eba6c0dde4d7cbb6d43b396298ce69041).